File: apps/web-ui/lib/due-risk.ts
export const PLANT_TIMEZONE = "Asia/Kolkata"
export const PRIORITY_WINDOW_DAYS = 3
export const DUE_RISK_PRIORITY = "PRIORITY"
export const DUE_RISK_OVERDUE = "OVERDUE"

export function plantToday(now = new Date()): string {
  return now.toLocaleDateString("en-CA", { timeZone: PLANT_TIMEZONE })
}

export function addPlantDays(isoDate: string, days: number): string {
  const [year, month, day] = isoDate.split("-").map(Number)
  const utc = Date.UTC(year, month - 1, day + days)
  const next = new Date(utc)
  const y = next.getUTCFullYear()
  const m = String(next.getUTCMonth() + 1).padStart(2, "0")
  const d = String(next.getUTCDate()).padStart(2, "0")
  return `${y}-${m}-${d}`
}

export function priorityWindow(today = plantToday()) {
  return { start: today, end: addPlantDays(today, PRIORITY_WINDOW_DAYS - 1) }
}

export function classifyDueRisk(dueDate?: string | null, today = plantToday()): string | null {
  if (!dueDate) return null
  const due = String(dueDate).slice(0, 10)
  if (!/^\d{4}-\d{2}-\d{2}$/.test(due)) return null
  if (due < today) return DUE_RISK_OVERDUE
  const { end } = priorityWindow(today)
  if (due <= end) return DUE_RISK_PRIORITY
  return null
}

export function dueRiskLabel(today = plantToday()) {
  const { start, end } = priorityWindow(today)
  const format = (value: string) => {
    const parsed = new Date(`${value}T00:00:00+05:30`)
    return parsed.toLocaleDateString("en-GB", { day: "2-digit", month: "short", timeZone: PLANT_TIMEZONE })
  }
  return `Priority delivery — next 3 plant days (${format(start)}–${format(end)}, ${PLANT_TIMEZONE})`
}

export function overdueLabel() {
  return "Overdue (due before today, plant calendar)"
}

File: hariom-erp/services/production-service/src/due_risk.py
"""One plant-local 3-day priority predicate used by tiles, lists, reports, and exports.

Priority delivery covers three calendar dates in Asia/Kolkata: today, today+1, and
today+2. Overdue (due_date < today) is a separate bucket and is never mixed into
priority. Jobs without a due date are unclassified.
"""

from __future__ import annotations

from datetime import date, datetime, timedelta
from typing import Optional
from zoneinfo import ZoneInfo

PLANT_TIMEZONE_NAME = "Asia/Kolkata"
PLANT_TIMEZONE = ZoneInfo(PLANT_TIMEZONE_NAME)
PRIORITY_WINDOW_DAYS = 3
DUE_RISK_PRIORITY = "PRIORITY"
DUE_RISK_OVERDUE = "OVERDUE"
OPEN_JOB_STATUSES_EXCLUDED = {"COMPLETED", "CANCELLED"}


def plant_today(now: Optional[datetime] = None) -> date:
    current = now or datetime.now(PLANT_TIMEZONE)
    if current.tzinfo is None:
        current = current.replace(tzinfo=ZoneInfo("UTC")).astimezone(PLANT_TIMEZONE)
    else:
        current = current.astimezone(PLANT_TIMEZONE)
    return current.date()


def priority_window(today: Optional[date] = None) -> tuple[date, date]:
    start = today or plant_today()
    return start, start + timedelta(days=PRIORITY_WINDOW_DAYS - 1)


def classify_due_risk(due_date: Optional[date], today: Optional[date] = None) -> Optional[str]:
    if due_date is None:
        return None
    current = today or plant_today()
    if due_date < current:
        return DUE_RISK_OVERDUE
    _start, window_end = priority_window(current)
    if due_date <= window_end:
        return DUE_RISK_PRIORITY
    return None


def is_open_job_status(status: Optional[str]) -> bool:
    return str(status or "").strip().upper() not in OPEN_JOB_STATUSES_EXCLUDED


def due_risk_label(today: Optional[date] = None) -> str:
    start, window_end = priority_window(today)
    return (
        f"Priority delivery — next 3 plant days "
        f"({start.strftime('%d %b')}–{window_end.strftime('%d %b')}, {PLANT_TIMEZONE_NAME})"
    )


def overdue_label() -> str:
    return "Overdue (due before today, plant calendar)"
